tests/sentry/sentry_metrics/test_gen_metrics_multiprocess_steps.py

```diff
@@ -0,0 +1,618 @@
+    )
+    message1 = Message(
+        BrokerValue(
+            KafkaPayload(None, b"some value", []), Partition(Topic("topic"), 0), 1, datetime.now()
+        )
+    )
+    batch_builder_time.append(message1)
+
+
+ts = int(datetime.now(tz=timezone.utc).timestamp())
+counter_payloads: list[dict[str, Any]] = [
+    {
+        "name": f"c:{use_case.value}/alert@none",
+        "tags": {
+            "environment": "production",
+            "session.status": "init",
+        },
+        "timestamp": ts,
+        "type": b"c",
+        "value": 1.0,
+        "org_id": 1,
+        "project_id": 3,
+        "retention_days": 90,
+    }
+    for use_case in UseCaseID
+    if use_case is not UseCaseID.SESSIONS
+]
+distribution_payloads: list[dict[str, Any]] = [
+    {
+        "name": f"d:{use_case.value}/alert@none",
+        "tags": {
+            "environment": "production",
+            "session.status": "healthy",
+        },
+        "timestamp": ts,
+        "type": b"d",
+        "value": [4, 5, 6],
+        "org_id": 1,
+        "project_id": 3,
+        "retention_days": 90,
+    }
+    for use_case in UseCaseID
+    if use_case is not UseCaseID.SESSIONS
+]
+
+set_payloads: list[dict[str, Any]] = [
+    {
+        "name": f"s:{use_case.value}/alert@none",
+        "tags": {
+            "environment": "production",
+            "session.status": "errored",
+        },
+        "timestamp": ts,
+        "type": b"s",
+        "value": [3],
+        "org_id": 1,
+        "project_id": 3,
+        "retention_days": 90,
+    }
+    for use_case in UseCaseID
+    if use_case is not UseCaseID.SESSIONS
+]
+
+
+def __translated_payload(
+    payload, indexer=None
+) -> Dict[str, Union[str, int, List[int], MutableMapping[int, int]]]:
+    """
+    Translates strings to ints using the MockIndexer
+    in addition to adding the retention_days
+
+    """
+    indexer = indexer or MESSAGE_PROCESSOR._indexer
+    MRI_RE_PATTERN = re.compile("^([c|s|d|g|e]):([a-zA-Z0-9_]+)/.*$")
+
+    payload = deepcopy(payload)
+    org_id = payload["org_id"]
+    matched_mri = MRI_RE_PATTERN.match(payload["name"])
+    assert matched_mri is not None
+    use_case_id = UseCaseID(matched_mri.group(2))
+
+    new_tags = {
+        indexer.resolve(use_case_id=use_case_id, org_id=org_id, string=k): v
+        for k, v in payload["tags"].items()
+    }
+    payload["metric_id"] = indexer.resolve(
+        use_case_id=use_case_id, org_id=org_id, string=payload["name"]
+    )
+    payload["retention_days"] = 90
+    payload["tags"] = new_tags
+    payload["use_case_id"] = use_case_id.value
+    payload["sentry_received_timestamp"] = BROKER_TIMESTAMP.timestamp()
+    payload["version"] = 2
+
+    payload.pop("unit", None)
+    del payload["name"]
+    return payload
+
+
+@pytest.mark.django_db
+def test_process_messages() -> None:
+    message_payloads = counter_payloads + distribution_payloads + set_payloads
+    message_batch: List[Any] = [
```

Comment:
this should probably be `list[Message]` I would imagine? -- then you don't need the `Any` below this

Reply:
This is what I tried initially, I got this from myPy
```
tests/sentry/sentry_metrics/test_gen_metrics_multiprocess_steps.py:346: error: "BaseValue[Any]" has no attribute "partition"  [attr-defined]
tests/sentry/sentry_metrics/test_gen_metrics_multiprocess_steps.py:347: error: "BaseValue[Any]" has no attribute "offset"  [attr-defined]
tests/sentry/sentry_metrics/test_gen_metrics_multiprocess_steps.py:348: error: "BaseValue[Any]" has no attribute "timestamp"  [attr-defined]
```
I also tried
```python
message_batch: list[Message[BrokerValue]] = [
```
Seeing that message is defined as `Message(Generic[TMessagePayload])` and value is annotated as `value: BaseValue[TMessagePayload]`
But I got
```
tests/sentry/sentry_metrics/test_gen_metrics_multiprocess_steps.py:322: error: Argument 1 to "BrokerValue" has incompatible type "KafkaPayload"; expected "BrokerValue[Any]"  [arg-type]
tests/sentry/sentry_metrics/test_gen_metrics_multiprocess_steps.py:335: error: Argument "outer_message" to "process_messages" of "MessageProcessor" has incompatible type "Message[List[Message[BrokerValue[Any]]]]"; expected "Message[List[Message[KafkaPayload]]]"  [arg-type]
tests/sentry/sentry_metrics/test_gen_metrics_multiprocess_steps.py:346: error: "BaseValue[BrokerValue[Any]]" has no attribute "partition"  [attr-defined]
tests/sentry/sentry_metrics/test_gen_metrics_multiprocess_steps.py:347: error: "BaseValue[BrokerValue[Any]]" has no attribute "offset"  [attr-defined]
tests/sentry/sentry_metrics/test_gen_metrics_multiprocess_steps.py:348: error: "BaseValue[BrokerValue[Any]]" has no attribute "timestamp"  [attr-defined]
```

Reply:
I also tried
```python
for m in message_batch:
        assert isinstance(m.value, BrokerValue)
```
Similarly to what I did [here](https://github.com/getsentry/sentry/pull/55950#discussion_r1327875828), but this type checker does not seem to be smart enough to understand the assertion in a for loop, so I am still getting

```
tests/sentry/sentry_metrics/test_rh_metrics_multiprocess_steps.py:331: error: "BaseValue[KafkaPayload]" has no attribute "partition"  [attr-defined]
tests/sentry/sentry_metrics/test_rh_metrics_multiprocess_steps.py:332: error: "BaseValue[KafkaPayload]" has no attribute "offset"  [attr-defined]
tests/sentry/sentry_metrics/test_rh_metrics_multiprocess_steps.py:333: error: "BaseValue[KafkaPayload]" has no attribute "timestamp"  [attr-defined]
```